### PR TITLE
Replace cudf::strings::detail::modify_strings utility with make_strings_children

### LINKED ISSUE #6649: [BUG] CSV writer uses wrong memory resource for temporary allocations
The memory allocations for internal temporary allocations should not be the one passed in to the external libcudf API.

#### in CSV writer, the memory resource passed to the external API here:
https://github.com/rapidsai/cudf/blob/60d25ea04bf018905836310a18133806f780cb22/cpp/src/io/functions.cpp#L183-L187
#### is stored in the impl class here:
https://github.com/rapidsai/cudf/blob/e25d1ca9d233036c5fce1fce81a2fb8823409e00/cpp/src/io/csv/writer_impl.cu#L365-L368
#### which constructs an internal functor and passes the same mr to it here:
https://github.com/rapidsai/cudf/blob/e25d1ca9d233036c5fce1fce81a2fb8823409e00/cpp/src/io/csv/writer_impl.cu#L518
#### and uses it for temporary allocations here:
https://github.com/rapidsai/cudf/blob/e25d1ca9d233036c5fce1fce81a2fb8823409e00/cpp/src/io/csv/writer_impl.cu#L242-L243 https://github.com/rapidsai/cudf/blob/e25d1ca9d233036c5fce1fce81a2fb8823409e00/cpp/src/io/csv/writer_impl.cu#L284 https://github.com/rapidsai/cudf/blob/e25d1ca9d233036c5fce1fce81a2fb8823409e00/cpp/src/io/csv/writer_impl.cu#L295 https://github.com/rapidsai/cudf/blob/e25d1ca9d233036c5fce1fce81a2fb8823409e00/cpp/src/io/csv/writer_impl.cu#L336

#### Additionally, the internal calls to other libcudf APIs should use the corresponding detail version and should pass the stream.

### PR DESCRIPTION
**Pull request by davidwendt**
The `cudf::strings::detail::modify_strings()` function is similar to `cudf::strings::detail::make_strings_children` but less efficient. The `modify_strings()` is only used by `cudf::strings::capitalize()` and the CSV writer. This PR replaces it with `make_strings_children`. This simplifies the code and also removes the need for header file `modify_strings.cuh`.

The CSV writer changes include fixes for #6649 to use the appropriate `detail` APIs and the default memory resource for temporary column data.
Closes #6649

### CODE REVIEW

**Comment by davidwendt** [comment]
rerun tests

File: cpp/src/io/csv/writer_impl.cu [hunk 1]
```diff
@@ -246,27 +155,20 @@ struct column_to_strings_fn {
...
-    //        str_column_modified = modify(str_row);
-    // where modify() = duplicate the double quotes, if any; add 2bl quotes prefix/suffix;
-    //}
-    //
     string_scalar delimiter{std::string{options_.get_inter_column_delimiter()}, true, stream_};
-    predicate_special_chars pred{delimiter.value(stream_)};
 
-    return modify_strings<probe_special_chars, modify_special_chars>(column_v, stream_, mr_, pred);
+    auto d_column = column_device_view::create(column_v, stream_);
+    escape_strings_fn fn{*d_column, delimiter.value()};
```
Comment (karthikeyann) [inline cpp/src/io/csv/writer_impl.cu:162]:
```suggestion
    escape_strings_fn fn{*d_column, delimiter.value(stream_)};
```

**Review by karthikeyann — APPROVED** [review]
Looks good to me. Nice.
(only 1 suggestion on stream)

**Review by cwharris — APPROVED** [review]

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**